File: apps/rivethub-web/src/lib/tool-titles.ts
/**
 * Human-readable titles for harness tool names (Claude Code + Grok Build).
 * Behavior ported from Android ChatMessageTools (display-only; not the APK).
 */

export type ToolArgs = Record<string, unknown> | undefined

function str(args: ToolArgs, key: string): string | undefined {
  if (!args) return undefined
  const v = args[key]
  return typeof v === 'string' && v.trim() ? v : undefined
}

function basename(path: string | undefined): string | undefined {
  if (!path) return undefined
  const parts = path.split(/[/\\]/)
  return parts[parts.length - 1] || path
}

/** Strip emoji / status prefixes from stream content ("🔧 shell" → "shell"). */
export function normalizeToolName(raw: string): string {
  const t = raw.replace(/^[\u{1F300}-\u{1FAFF}\u{2600}-\u{27BF}\s✅❌🔧]+/u, '').trim()
  return t || raw.trim() || 'tool'
}

/**
 * Human title for a tool invocation. Unknown names fall back to a short form
 * of the raw name (not empty).
 */
export function humanToolTitle(rawName: string, args?: ToolArgs): string {
  const name = normalizeToolName(rawName)
  const lower = name.toLowerCase()

  // Claude Code
  if (name === 'Bash' || lower === 'bash' || lower === 'shell') {
    const d = str(args, 'description') ?? str(args, 'command')?.slice(0, 48)
    return d ? `Ran: ${d}` : 'Ran a command'
  }
  if (name === 'Edit' || name === 'NotebookEdit') {
    return `Edited ${basename(str(args, 'file_path')) ?? 'file'}`
  }
  if (name === 'Write') {
    return `Wrote ${basename(str(args, 'file_path')) ?? 'file'}`
  }
  if (name === 'Read') {
    return `Read ${basename(str(args, 'file_path')) ?? 'file'}`
  }
  if (name === 'Grep' || name === 'Glob') {
    const p = str(args, 'pattern') ?? ''
    return p ? `Searched: ${p}` : 'Searched files'
  }
  if (name === 'WebFetch') {
    return `Fetched ${str(args, 'url') ?? 'page'}`
  }
  if (name === 'WebSearch') {
    return `Searched web: ${str(args, 'query') ?? ''}`
  }
  if (name === 'Task') {
    return `Subagent: ${str(args, 'description') ?? 'task'}`
  }
  if (name === 'AskUserQuestion') {
    return 'Asked a question'
  }

  // Grok Build / common agent tools
  if (lower === 'run_terminal_command' || lower === 'run_terminal_cmd') {
    const d = str(args, 'description') ?? str(args, 'command')?.slice(0, 48)
    return d ? `Ran: ${d}` : 'Ran a command'
  }
  if (lower === 'read_file') {
    return `Read ${basename(str(args, 'path') ?? str(args, 'file_path')) ?? 'file'}`
  }
  if (lower === 'search_replace' || lower === 'edit_file' || lower === 'apply_patch') {
    return `Edited ${basename(str(args, 'path') ?? str(args, 'file_path')) ?? 'file'}`
  }
  if (lower === 'write_file' || lower === 'create_file') {
    return `Wrote ${basename(str(args, 'path') ?? str(args, 'file_path')) ?? 'file'}`
  }
  if (lower === 'grep' || lower === 'glob' || lower === 'find_files' || lower === 'list_dir') {
    const p = str(args, 'pattern') ?? str(args, 'query') ?? str(args, 'path') ?? ''
    return p ? `Searched: ${p}` : 'Searched files'
  }
  if (lower === 'web_search') {
    return `Searched web: ${str(args, 'query') ?? ''}`
  }
  if (lower === 'web_fetch') {
    return `Fetched ${str(args, 'url') ?? 'page'}`
  }
  if (lower === 'todo_write') {
    return 'Updated task list'
  }
  if (lower === 'ask_user_question' || lower === 'ask_user') {
    return 'Asked a question'
  }

  // MCP-style "mcp:server:tool" → last segment
  if (name.includes(':')) {
    const last = name.split(':').pop() ?? name
    return last.replace(/_/g, ' ')
  }

  return name.replace(/_/g, ' ')
}

File: apps/rivethub-android/app/src/main/java/io/rivethub/app/plane/ToolTitles.kt
package io.rivethub.app.plane

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull

/**
 * Human-readable titles for harness tool names (Claude Code + Grok Build).
 * Port of rivethub-web `lib/tool-titles.ts`.
 */

fun toolArgStrings(args: JsonObject?): Map<String, String> {
    if (args == null) return emptyMap()
    val out = LinkedHashMap<String, String>()
    for ((k, v) in args) {
        val p = v as? JsonPrimitive ?: continue
        if (!p.isString) continue
        val s = p.contentOrNull?.trim().orEmpty()
        if (s.isNotEmpty()) out[k] = s
    }
    return out
}

// normalizeToolName lives in AskUser.kt (same package) — one definition.

private fun arg(args: Map<String, String>, key: String): String? =
    args[key]?.takeIf { it.isNotBlank() }

private fun basename(path: String?): String? {
    if (path.isNullOrEmpty()) return null
    val parts = path.split('/', '\\')
    return parts.lastOrNull()?.ifEmpty { path } ?: path
}

fun humanToolTitle(rawName: String, args: Map<String, String> = emptyMap()): String {
    val name = normalizeToolName(rawName)
    val lower = name.lowercase()

    if (name == "Bash" || lower == "bash" || lower == "shell") {
        val d = arg(args, "description") ?: arg(args, "command")?.take(48)
        return if (d != null) "Ran: $d" else "Ran a command"
    }
    if (name == "Edit" || name == "NotebookEdit") {
        return "Edited ${basename(arg(args, "file_path")) ?: "file"}"
    }
    if (name == "Write") {
        return "Wrote ${basename(arg(args, "file_path")) ?: "file"}"
    }
    if (name == "Read") {
        return "Read ${basename(arg(args, "file_path")) ?: "file"}"
    }
    if (name == "Grep" || name == "Glob") {
        val p = arg(args, "pattern").orEmpty()
        return if (p.isNotEmpty()) "Searched: $p" else "Searched files"
    }
    if (name == "WebFetch") {
        return "Fetched ${arg(args, "url") ?: "page"}"
    }
    if (name == "WebSearch") {
        return "Searched web: ${arg(args, "query").orEmpty()}"
    }
    if (name == "Task") {
        return "Subagent: ${arg(args, "description") ?: "task"}"
    }
    if (name == "AskUserQuestion") {
        return "Asked a question"
    }

    if (lower == "run_terminal_command" || lower == "run_terminal_cmd") {
        val d = arg(args, "description") ?: arg(args, "command")?.take(48)
        return if (d != null) "Ran: $d" else "Ran a command"
    }
    if (lower == "read_file") {
        return "Read ${basename(arg(args, "path") ?: arg(args, "file_path")) ?: "file"}"
    }
    if (lower == "search_replace" || lower == "edit_file" || lower == "apply_patch") {
        return "Edited ${basename(arg(args, "path") ?: arg(args, "file_path")) ?: "file"}"
    }
    if (lower == "write_file" || lower == "create_file") {
        return "Wrote ${basename(arg(args, "path") ?: arg(args, "file_path")) ?: "file"}"
    }
    if (lower == "grep" || lower == "glob" || lower == "find_files" || lower == "list_dir") {
        val p = arg(args, "pattern") ?: arg(args, "query") ?: arg(args, "path").orEmpty()
        return if (p.isNotEmpty()) "Searched: $p" else "Searched files"
    }
    if (lower == "web_search") {
        return "Searched web: ${arg(args, "query").orEmpty()}"
    }
    if (lower == "web_fetch") {
        return "Fetched ${arg(args, "url") ?: "page"}"
    }
    if (lower == "todo_write") {
        return "Updated task list"
    }
    if (lower == "ask_user_question" || lower == "ask_user") {
        return "Asked a question"
    }

    if (name.contains(':')) {
        val last = name.substringAfterLast(':')
        return last.replace('_', ' ')
    }
    if (name.contains("__")) {
        val last = name.substringAfterLast("__")
        return last.replace('_', ' ')
    }
    return name.replace('_', ' ')
}
